File: examples/nodejs/19-openai-compatible.ts
/**
 * OpenAI-Compatible Provider Example
 *
 * Demonstrates using OpenRouter through the generic OpenAI-compatible
 * provider interface. This shows that any OpenAI-compatible API can be
 * used with LLMKit.
 *
 * Note: The dedicated OpenRouterProvider is recommended for Qwen3 models
 * because it automatically handles the `/no_think` injection. When using
 * the OpenAI-compatible endpoint, you need to manually add `/no_think`
 * to prompts or use higher max_tokens.
 *
 * Requirements:
 * - Set OPENROUTER_API_KEY environment variable
 * - npm install
 *
 * Run:
 *   npx ts-node 18-openai-compatible.ts
 */

import {
    JsLLMKitClient as LLMKitClient,
    JsMessage as Message,
    JsCompletionRequest as CompletionRequest,
} from 'llmkit'

async function main() {
    const apiKey = process.env.OPENROUTER_API_KEY
    if (!apiKey) {
        throw new Error('OPENROUTER_API_KEY environment variable not set')
    }

    // Create client with OpenAI-compatible provider pointing to OpenRouter
    // Note: Provider name (modelId) cannot contain dashes due to model format parsing
    const client = new LLMKitClient({
        providers: {
            openaiCompatible: {
                apiKey: apiKey,
                baseUrl: 'https://openrouter.ai/api/v1',
                modelId: 'openroutercompat',  // Provider name for routing
            }
        }
    })

    // Model format is "provider/model" - everything after first "/" is the model name
    const model = 'openroutercompat/qwen/qwen3-32b'

    console.log('OpenAI-Compatible Provider Example')
    console.log('='.repeat(50))
    console.log('Using OpenRouter via OpenAI-compatible endpoint\n')

    // Since we're not using OpenRouterProvider, we need to manually
    // add /no_think to disable Qwen3's reasoning mode
    console.log('Request with manual /no_think:')
    console.log('-'.repeat(40))

    let request = CompletionRequest
        .create(model, [Message.user('What is 2 + 2? Answer briefly. /no_think')])
        .withSystem('You are a helpful assistant.')
        .withMaxTokens(100)

    let response = await client.complete(request)
    console.log(`Response: ${response.textContent().trim()}\n`)

    // Second request
    console.log('Another request:')
    console.log('-'.repeat(40))

    request = CompletionRequest
        .create(model, [Message.user('Name 3 colors. /no_think')])
        .withSystem('Be concise.')
        .withMaxTokens(100)

    response = await client.complete(request)
    console.log(`Response: ${response.textContent().trim()}\n`)

    console.log('='.repeat(50))
    console.log('OpenRouter works via OpenAI-compatible endpoint!')
    console.log('Note: Manual /no_think required for Qwen3 models.')
}

main().catch(console.error)

File: examples/python/19_openai_compatible.py
"""
OpenAI-Compatible Provider Example

Demonstrates using OpenRouter through the generic OpenAI-compatible
provider interface. This shows that any OpenAI-compatible API can be
used with LLMKit.

Note: The dedicated OpenRouterProvider is recommended for Qwen3 models
because it automatically handles the `/no_think` injection. When using
the OpenAI-compatible endpoint, you need to manually add `/no_think`
to prompts or use higher max_tokens.

Requirements:
- Set OPENROUTER_API_KEY environment variable
- uv sync

Run:
    uv run python 19_openai_compatible.py
"""

import os
from llmkit import LLMKitClient, Message, CompletionRequest


def main():
    api_key = os.environ.get("OPENROUTER_API_KEY")
    if not api_key:
        raise ValueError("OPENROUTER_API_KEY environment variable not set")

    # Create client with OpenAI-compatible provider pointing to OpenRouter
    # Note: Provider name (model_id) cannot contain dashes due to model format parsing
    client = LLMKitClient(
        providers={
            "openai_compatible": {
                "api_key": api_key,
                "base_url": "https://openrouter.ai/api/v1",
                "model_id": "openroutercompat",  # Provider name for routing
            }
        }
    )

    # Model format is "provider/model" - everything after first "/" is the model name
    model = "openroutercompat/qwen/qwen3-32b"

    print("OpenAI-Compatible Provider Example")
    print("=" * 50)
    print("Using OpenRouter via OpenAI-compatible endpoint\n")

    # Since we're not using OpenRouterProvider, we need to manually
    # add /no_think to disable Qwen3's reasoning mode
    print("Request with manual /no_think:")
    print("-" * 40)

    request = CompletionRequest(
        model=model,
        messages=[Message.user("What is 2 + 2? Answer briefly. /no_think")],
        system="You are a helpful assistant.",
        max_tokens=100,
    )

    response = client.complete(request)
    print(f"Response: {response.text_content().strip()}\n")

    # Second request
    print("Another request:")
    print("-" * 40)

    request = CompletionRequest(
        model=model,
        messages=[Message.user("Name 3 colors. /no_think")],
        system="Be concise.",
        max_tokens=100,
    )

    response = client.complete(request)
    print(f"Response: {response.text_content().strip()}\n")

    print("=" * 50)
    print("OpenRouter works via OpenAI-compatible endpoint!")
    print("Note: Manual /no_think required for Qwen3 models.")


if __name__ == "__main__":
    main()
